pubsub/index.js

```diff
@@ -68,35 +74,46 @@ const extractInfoFromMessage = (message) => {
   };
 };
 
-// Get attachment of a message..
+// Get attachment of a message.
 const extractAttachmentFromMessage = async (email, messageId, attachmentId) => {
-  await gmail.users.messages.attachments.get({
+  return gmail.users.messages.attachments.get({
     id: attachmentId,
     messageId: messageId,
     userId: email
   });
 };
 
-// Upload attachment of a message to Cloud Storage.
-const uploadAttachment = async (data, filename) => {
-  let file = storageClient.bucket(BUCKET).file(filename);
-  let writeableStream = file.createWriteStream({resumable: false});
-  writeableStream.write(data);
-  writeableStream.end();
+// Tag the attachment using Cloud Vision API
+const analyzeAttachment = async (data, filename) => {
+  var topLabels = ['', '', ''];
+  if (filename.endsWith('.png') || filename.endsWith('.jpg')) {
+    const [analysis] = await visionClient.labelDetection({
+      image: {
+        content: Buffer.from(data, 'base64')
+      }
+    });
+    const labels = analysis.labelAnnotations;
+    for (var i = 0; i <= 2; i++) {
+      if (labels[i] && labels[i].description) {
+        topLabels[i] = labels[i].description;
+      }
+    }
+  }
+
+  return topLabels;
 };
 
 // Write sender, attachment filename, and download link to a Google Sheet.
-const updateReferenceSheet = async (from, filename) => {
-  let link = `https://storage.cloud.google.com/${BUCKET}/${filename}`;
+const updateReferenceSheet = async (from, filename, topLabels) => {
   await googleSheets.spreadsheets.values.append({
     spreadsheetId: SHEET,
-    range: 'Sheet1!A1:D1',
+    range: 'Sheet1!A1:F1',
```

Comment:
Are both of these `range` properties necessary?

If so - can we move this to a separate variable?

Reply:
Yeah, those ranges are required I think. I moved it to a separate variable.